Add public API for ILVerify

jcouv opened the pull request:
@ArztSamuel @jkotas @VSadov  Here is another stab at a public API for ILVerify.
The ILVerify program, the ILVerify tests and the Roslyn tests all rely on the same API, summarized below. There are two APIs that are internal and used by the ILVerify tests.
```C#
namespace ILVerify
{
    public delegate bool ShouldVerifyMethod(string name);

    public interface IResolver
    {
        PEReader Resolve(AssemblyName name);
    }

    public class Verifier
    {
        public ShouldVerifyMethod ShouldVerifyMethod { set; } // Used by ILVerify program to filter which methods should be verified vs. skipped
        public Verifier(IResolver resolver);
        internal Verifier(SimpleTypeSystemContext context); // Used by the ILVerify tests
        public void SetSystemModuleName(AssemblyName name);
        public VerificationResult Verify(AssemblyName moduleToVerify);
    }

    public class VerificationResult
    {
        public int NumErrors = 0;
        public string Message = string.Empty;
        internal IEnumerable<VerifierError> _errors; // Note: there may be fewer errors recorded here than counted in NumErrors, which also counts exceptions. Used by the ILVerify tests
    }

    public class VerifierException : Exception
    {
        public VerifierException(string message) : base(message)
        {
        }
    }
}
```

Note 1: there is auto-discovery feature for the system module. It is required for Roslyn tests which create their own core library, but the test framework doesn't know when that is the case. If this feature feels wrong, I can further explore a Roslyn-side solution.

Note 2: the `SimpleTypeSystemContext` keeps a cache of `EcmaModule`s (`_modules`). That is currently keyed on simple name instead of full name. That's because I don't know how to get a full name from an `EcmaModule` or `MetadataReader` (see `CreateModule` method). Any tips?

If the API looks good, I will clean up this PR (removing strong name using Roslyn key, switch back to an exe).

Fixes https://github.com/dotnet/corert/issues/3734

jkotas commented on src/Common/src/TypeSystem/Common/TypeSystemContext.cs:
```diff
@@ -66,6 +66,13 @@ public virtual ModuleDesc ResolveAssembly(AssemblyName name, bool throwIfNotFoun
             return null;
         }
 
+        public virtual ModuleDesc ResolveModule(AssemblyName name, bool throwIfNotFound = true)
```
If you want to add support for multi-module assemblies, it should be done in separate PR.

jkotas commented:
>  there is auto-discovery feature for the system module

It would be nice to submit this in separate PR as well.

jkotas commented:
> That's because I don't know how to get a full name from an `EcmaModule`

Modules in multimodule assemblies do not have assembly names. They are always resolved in the context of the owning assembly.

jkotas commented:
> There are two APIs that are internal and used by the ILVerify tests.

These are not the only internal APIs used by the ILVerify tests...

jcouv commented:
Updated the PR with Jan's feedback (removed the multi-module and auto-discovery features).
Any feedback on the public API design?

Also, I'm thinking the `exe` should be split into `exe`+`dll`. Jared mentioned that using `exe`'s as library often causes problems. And the split should happen before release packages get published.

jkotas commented:
> Any feedback on the public API design?

- ShouldVerifyMethod: I think it would be better to have a method that takes a metadata token and that you call to verify a single method.
- VerificationResult: Why is this not a simple array of errors?
- VerifierException: Does it need to be part of the public API?

> Also, I'm thinking the exe should be split into exe+dll

Agree that these should be split.

MichalStrehovsky commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,224 @@
+                    continue;
+                }
+
+                var methodName = method.ToString();
```
`ToString()` on the type system objects is a debugging helper, not an API. We reserve the right to change the output any time to aid debuggability. I recently overhauled it in #5065.

ArztSamuel commented:
I noticed that the tests were not executable any longer after these changes. This is caused by the method tester now resolving the test assemblies in a different way, which enforces assembly name == file name. Some older test assemblies did not respect this. I have submitted a PR fixing it: https://github.com/dotnet/corert/pull/5198

jkotas commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,224 @@
+            return new VerificationResult();
+        }
+
+        internal VerificationResult VerifyMethod(MethodDesc method, MethodIL methodIL, string moduleName)
```
Do these methods need to take `moduleName`? You should be always able to get the module name from `MethodDesc`.

jcouv replied:
I should rename the `moduleName`. This parameter is used to maintain the old command-line output, which inclues file paths.
But when the verifier is used in-memory, I'm passing in the module name.

ArztSamuel commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,224 @@
+            }
+            catch (VerificationException)
+            {
+                numErrors++;
```
`VerificationException`s are thrown to either abort the entire method or a single basic block. When such an event happens an error should already have been reported, therefore increasing `numErrors` here is redundant.
It also breaks some test cases.

jcouv replied:
Thanks.
It seems I've not been running all the tests locally and CI isn't either :-( Investing in test automation would surely help.

ArztSamuel commented:
> I think it would be better to have a method that takes a metadata token and that you call to verify a single method.

I agree. As discussed in the issue, it would make sense to me, to have:
`Verify(PEReader)` for verifying entire modules
`Verify(PEReader, MethodDefinitionHandle)` for verifying a specific method
`Verify(PEReader, TypeDefinitionHandle)` for verifying a specific type

As far as the auto-discovery is concerned: Would it make sense to add a property `SystemModuleName` with getter to `IResolver`? Or do you think this does not belong into the Resolver?
The ILVerify Program would simply return `_systemModule` for it, while a resolver implementation for the Roslyn tests could perform auto detection.

ArztSamuel commented on src/ILVerify/src/Program.cs:
```diff
@@ -3,42 +3,28 @@
-        private string _systemModule = DefaultSystemModuleName;
-        private Dictionary<string, string> _inputFilePaths = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
-        private Dictionary<string, string> _referenceFilePaths = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        private AssemblyName _systemModule;
```
Shouldn't this be set to `AssemblyName("mscorlib")` by default?

jcouv replied:
We don't know if the caller is using simple names or full names as identifiers.

jcouv replied:
Oops. You were right :-)

ArztSamuel commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,224 @@
+                var result = VerifyMethod(method, methodIL, path);
+                if (result.NumErrors > 0)
+                {
+                    return result;
```
Immediately returning here will abort the module verification after the first method has reported an error. This is different from how we and PEVerify previously handled this (keep verifying as long as sensible).

ArztSamuel commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,224 @@
+        public int NumErrors = 0;
+        public string Message = string.Empty;
+        internal IEnumerable<VerifierError> _errors; // Note: there may be fewer errors recorded here than counted in NumErrors, which also counts exceptions
+    }
```
I agree with @jkotas. I think it would make sense to simply have an enumeration of `VerificationError`s (which can be an array internally).
I think a new `VerificationError` class would be helpful for this, which could look like this:
```csharp
class VerificationError
{
    public readonly string ModuleName;
    public readonly string MethodName;
    public readonly VerificationErrorArgs ErrorArgs;
}
```
It might even make sense to use a `MethodDefinitionHandle` and `PEReader` instead of strings, to be more consistent with the proposed `Verify` method.
I am currently working on a use case, where having the VerificationErrorArgs for each reported error is vital. Creating the error string can then also be implemented by the application using the Verifier rather than the Verifier having to implement it.

jcouv replied:
I went with strings. Feel free to adjust to your need, after this PR but before we publish.

jcouv commented:
@jkotas 
> ShouldVerifyMethod: I think it would be better to have a method that takes a metadata token and that you call to verify a single method.

Just to confirm your proposal: the user of the API should be responsible for enumerating all method definitions that it cares about and verify them one by one (using a public API like `VerifyMethod(MethodDesc method, MethodIL methodIL)` rather than `Verify(AssemblyName moduleToVerify)` with a callback for filtering methods).

jkotas commented:
> the user of the API should be responsible for enumerating all method definitions that it cares about and verify them one by one

Yes - when the user wants to filter which methods to verify.


> sing a public API like VerifyMethod(MethodDesc method, MethodIL methodIL)

The public API should be: `void VerifyMethod(System.Reflection.PortableExecutable.PEReader peReader, System.Reflection.Metadata.MethodDefinitionHandle methodHandle)` as proposed in  https://github.com/dotnet/corert/pull/5186#issuecomment-355273478

MethodDesc is internal implementation detail. It should not show up in any public APIs. We should make sure that it is not marked as public in the final version of the .dll.

jcouv commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,224 @@
+            _typeSystemContext.SetSystemModule(_typeSystemContext.GetModule(name));
+        }
+
+        public VerificationResult Verify(AssemblyName moduleToVerify)
```
> I agree. As discussed in the corresponding issue, it would make sense to me, to have:
> Verify(PEReader) for verifying entire modules
> Verify(PEReader, MethodDefinitionHandle) for verifying a specific method
> Verify(PEReader, TypeDefinitionHandle) for verifying a specific type

If the API accepts `PEReader`, then we have to give up caching in `SimpleTypeSystemContext`, since `PEReader`s can't be keys in a dictionary. New `EcmaModule`s will be recreated multiple times.
I don't know if that is just a performance concern or if that might cause identity problems as well...

jkotas replied:
> since PEReaders can't be keys in a dictionary

Why not? We can use their object identity.

jcouv replied:
The `PEReader`s returned by `IResolver.Resolve` are currently not cached, so we're getting new ones. I'll add caching then object identity should work.

jcouv commented:
Pushed a commit that addresses the feedback so far.
I think this project needs some infrastructure for compiling IL in unittests. That will help testing the APIs. The core logic is already shared and therefore tested by existing tests.

MichalStrehovsky commented:
> I think this project needs some infrastructure for compiling IL in unittests

We have all the scaffolding there. In #3725 the ILVerify project made the jump to netcoreapp2.0 before the repo was able to deal with it, so the build and testing got disconnected from how the rest of the repo is built. As you note, we don't run the tests in the CI, and we don't even build the project in the CI.

At this point we should be able to [hook it up](https://github.com/dotnet/corert/blob/d939cca2706835138dbb6402e489e9e9f6f52f25/src/dirs.proj#L9-L11) into the build again and make whatever changes are necessary for it to build with `build.cmd` (in a separate pull request). A sample of a test that has assets in IL is [the typesystem test](https://github.com/dotnet/corert/tree/d939cca2706835138dbb6402e489e9e9f6f52f25/src/ILCompiler.TypeSystem/tests).

ArztSamuel commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,255 @@
+
+        private SimpleTypeSystemContext _typeSystemContext;
+
+        private static VerificationResult s_noSystemModuleResult = new VerificationResult() { Message = "No system module specified" };
```
Would it make sense to also throw an exception for this case instead?

ArztSamuel commented:
> It seems I've not been running all the tests locally and CI isn't either :-( Investing in test automation would surely help.

I totally agree. That's not your fault.
Automating the test execution is actually the next thing on my list for ILVerify. I have been assembling the tests with a cmd script locally for quite a while now, I just didn't have time to implement it properly yet.

jcouv commented:
@ArztSamuel Roslyn has some test infrastructure to compile IL on the fly (see "CompileIL"), if that helps. Takes a string as input (and maybe some references?) and produces an array of bytes.

jkotas commented on src/ILVerify/src/VerificationResult.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    public class VerificationResult
+    {
+        public string ModuleName { get; internal set; }
```
Do we need ModuleName here? The verification granularity is on module, so it will be always same for the given invocation.

If we need it here, it should be `PEReader` that the caller gets the ModuleName from if it needs it.

jkotas commented on src/ILVerify/src/VerificationResult.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        public string ModuleName { get; internal set; }
+        public string TypeName { get; internal set; }
+        public string Method { get; internal set; }
```
This should be MethodDefinitionHandle .

jkotas commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,255 @@
+
+        public IEnumerable<VerificationResult> Verify(PEReader peReader, TypeDefinitionHandle typeHandle)
+        {
+            if (peReader is null)
```
Some places use `is null`, other places use `== null`. It would be nice to be consistent. Personally, `is null` reads harder to me than `== null`.

jkotas commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,255 @@
+
+            if (_typeSystemContext.SystemModule is null)
+            {
+                yield return s_noSystemModuleResult;
```
I do not think this should be verification error. It should just throw exception - it is invalid use.

jcouv commented on src/ILVerify/src/SimpleTypeSystemContext.cs:
```diff
@@ -3,132 +3,82 @@
-
 namespace ILVerify
 {
     class SimpleTypeSystemContext : MetadataTypeSystemContext
```
Sure. I'll rename as last thing (and also split into exe+dll) once review is ok, so that the diff remains readable.

jkotas replied:
> and also split into exe+dll

I would do the split in separate PR.

jkotas commented on src/ILVerify/src/SimpleTypeSystemContext.cs:
```diff
@@ -3,118 +3,44 @@
-        {
-            // This is called once for every assembly that should be verified, so linear search is acceptable.
-            foreach (KeyValuePair<EcmaModule, ModuleData> entry in _moduleData)
+            string actualSimpleName = peReader.GetSimpleName();
```
This will create MetadataReader and throw it away immediately. The GetModule method a few lines below will create the MetadataReader again.

It would be better to move this validation to after the GetModule call (and delete the GetSimpleName method because of it should not be needed anymore).

jkotas commented on src/ILVerify/src/VerificationResult.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public class VerificationResult
+    {
+        public string TypeName { get; internal set; }
```
We should not need TypeName. It can be fetched from MethodDefinitionHandle.

jcouv replied:
Removed `TypeName`

jkotas commented on src/ILVerify/src/Program.cs:
```diff
@@ -97,169 +85,146 @@ private ArgumentSyntax ParseCommandLine(string[] args)
+            Write(" : ");
+            Write(result.TypeName);
+            Write("::");
+            Write(result.Method);
```
I do not think this will print a nice name of the method.

jcouv replied:
Thanks
Fixed this and verified manually. Here's an example of output:
```
[IL]: Error: [D:\issues\Console46\Console46\bin\Debug\New.exe : Program::set_Property([mscorlib]System.Int32)][offset 0x00000002] Fall through end of the method without returning.
1 Error(s) Verifying D:\issues\Console46\Console46\bin\Debug\New.exe
```

jkotas commented:
LGTM modulo comments

ArztSamuel commented on src/ILVerify/src/Program.cs:
```diff
@@ -97,169 +85,146 @@ private ArgumentSyntax ParseCommandLine(string[] args)
+                }
 
-                    message.Append(" ");
+                var numErrors = results.Count();
```
I think since `VerifyAssembly` is implemented using `yield return`, this will verify the entire assembly a second time, just to get the error count. It might make sense to simply increase a counter in the above foreach instead.

jcouv replied:
Definitely. Thanks!

jkotas commented on src/Common/src/TypeSystem/Ecma/MetadataExtensions.cs:
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
 using System.Reflection;
 using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
```
This line should not be necessary anymore.

jkotas commented on src/ILVerify/src/Verifier.cs:
```diff
@@ -0,0 +1,234 @@
+            _typeSystemContext.SetSystemModule(_typeSystemContext.GetModule(_typeSystemContext._resolver.Resolve(name)));
+        }
+
+        public EcmaModule GetModule(PEReader peReader)
```
`EcmaModule` should not referenced in the public surface of the eventual ILVerify library. Could you please make it internal?

jkotas commented:
Could please fix the few remaining nits, and undo the strongname signing hacks? I think we can merge this, and then start new PR for the additional changes.

jcouv commented:
Cool. Thanks a lot for the review and feedback.
I'll go ahead with a PR to split to exe+dll. 

Also tagging @A-And as FYI since he's working on nuget package.
